Add usage description to readme (#50)
yiisoft/user: README.md

## General usage

This package is an addition to [yiisoft/auth](https://github.com/yiisoft/auth)
and provides additional functionality for interacting with user identity.

### Working with identity

> The `CurrentUser` class is responsible for login and logout, as well as receiving data about the current user.

Reviewer: ```suggestion
The `CurrentUser` class is responsible for login and logout, as well as for providing data about the current user identity.
```

$currentUser->getIdentity(); // Original identity instance
```

This can be useful to check the user's problems during administration.

#### Login and logout

Methods of the same name are provided for login and logout:

```php
/**
 * @var \Yiisoft\Auth\IdentityInterface $identity
 */

$currentUser->getIdentity(); // GuestIdentityInterface instance

if ($currentUser->login($identity)) {
    $currentUser->getIdentity(); // $identity
    // Some actions
}

if ($currentUser->logout()) {
    $currentUser->getIdentity(); // GuestIdentityInterface instance
    // Some actions
}
```

In the process of login and logout triggers several events. Events are classes:

- `Yiisoft\User\Event\BeforeLogin` - triggered at the beginning of login process.
  Listeners of this event may call `$event->invalidate()` to cancel the login process.
- `Yiisoft\User\Event\AfterLogin` - triggered at the ending of login process.
- `Yiisoft\User\Event\BeforeLogout` - triggered at the beginning of logout process.
  Listeners of this event may call `$event->invalidate()` to cancel the logout process.
- `Yiisoft\User\Event\AfterLogout` - triggered at the ending of logout process.

> Listeners of all these events can get an identity instance participating in the process using `$event->getIdentity()`.
> These events are passed to the `Psr\EventDispatcher\EventDispatcherInterface` implementation, which is specified in the
> constructor when the `Yiisoft\User\CurrentUser` instance is initialized.

Reviewer: ```suggestion
Listeners of these events can get an identity instance participating in the process using `$event->getIdentity()`.
Events are dispatched by `Psr\EventDispatcher\EventDispatcherInterface` instance, which is specified in the
constructor when the `Yiisoft\User\CurrentUser` instance is initialized.
```

#### Checking user access

> To check whether the user can perform the operation as specified by the given permission,
> you need to set an access checker (see [yiisoft/access](https://github.com/yiisoft/access)) instance:

Reviewer: ```suggestion
To be able to check whether the current user can perform an operation corresponding to a given permission,
you need to set an access checker (see [yiisoft/access](https://github.com/yiisoft/access)) instance:
```

```php
/** 
 * @var \Yiisoft\Access\AccessCheckerInterface $accessChecker
 */
 
$currentUser = $currentUser->withAccessChecker($accessChecker);
```

> And to check, you need to use the `can()` method:

Reviewer: ```suggestion
To perform the check, use `can()` method:
```

```php
// The name of the permission (e.g. "edit post") that needs access check.
$permissionName = 'edit-post'; // Required.

// Name-value pairs that would be passed to the rules associated with the roles and permissions assigned to the user.
$params = ['postId' => 42]; // Optional. Default is empty array.

if ($currentUser->can($permissionName, $params)) {
    // Some actions
}
```

> Note that you must first provide access checker via `withAccessChecker()` method.
> Otherwise, the `can()` method will always return `false`.

Reviewer: ```suggestion
Note that in case access checker is not provided via `withAccessChecker()` method,
`can()` will always return `false`.
```

#### Session usage

> The current user can store in the session the necessary user ID and auth timeouts for auto login.
> To do this, you need to set a session (see [yiisoft/session](https://github.com/yiisoft/session)) instance:

Reviewer: ```suggestion
The current user can store user ID and authentication timeouts for auto-login in the session.
To do this, you need to provide a session (see [yiisoft/session](https://github.com/yiisoft/session)) instance:
```

```php
/** 
 * @var \Yiisoft\Session\SessionInterface $session
 */
 
$currentUser = $currentUser->withSession($session);
```

You can set timeout (number of seconds), during which the user will be logged
out automatically in case of remaining inactive:

```php
$currentUser = $currentUser->withAuthTimeout(3600);
```

As well as an absolute timeout (number of seconds), during which the user will be logged
out automatically regardless of activity:

```php
$currentUser = $currentUser->withAbsoluteAuthTimeout(3600);
```

By default, timeouts are not used, the user will be logged out after the current session expires.

#### Using with event loop

> The `Yiisoft\User\CurrentUser` instance are stateful, so when you build long-running applications
> with tools like [Swoole](https://www.swoole.co.uk/) or [RoadRunner](https://roadrunner.dev/) you should reset
> the state at every request. For this purpose, you can use the `clear()` method.

Reviewer: ```suggestion
The `Yiisoft\User\CurrentUser` instance is stateful, so when you build long-running applications
with tools like [Swoole](https://www.swoole.co.uk/) or [RoadRunner](https://roadrunner.dev/) you should reset
the state at every request. For this purpose, you can use the `clear()` method.
```

### Auto login through identity from request attribute

When using authentication methods and an authentication middleware that authenticates
the user and places an `Yiisoft\Auth\IdentityInterface` instance in the corresponding
request attribute(see [yiisoft/auth](https://github.com/yiisoft/auth)).
For auto login, use `Yiisoft\User\Login\LoginMiddleware`.

> Please note that is mandatory before `Yiisoft\User\Login\LoginMiddleware`,

> > in the middleware stack should be located `Yiisoft\Auth\Middleware\Authentication`.

Reviewer: Should be corrected. Current text doesn't reflect the meaning.

### Auto login through cookie

In order to log user in automatically based on request cookie presence,
use `Yiisoft\User\Login\Cookie\CookieLoginMiddleware`.

To use a middleware, you need to implement `Yiisoft\User\Login\Cookie\CookieLoginIdentityInterface`
and also implement and use an instance of `IdentityRepositoryInterface` in the `CurrentUser`,
which will return `CookieLoginIdentityInterface`:

```php
use App\CookieLoginIdentity;
use Yiisoft\Auth\IdentityRepositoryInterface;

final class CookieLoginIdentityRepository implements IdentityRepositoryInterface
{
    private Storage $storage;

    public function __construct(Storage $storage)
    {
        $this->storage = $storage;
    }

    public function findIdentity(string $id): ?Identity
    {   
        return new CookieLoginIdentity($this->storage->findOne($id));
    }
}
```

> The login process itself when requested, the `CookieLoginMiddleware` will do automatically.

Reviewer: Should be re-phrased.

#### Creating a cookie

By default, you should set cookie for auto login manually in your application after logging user in:

```php
public function login(
    \Psr\Http\Message\ServerRequestInterface $request,
    \Psr\Http\Message\ResponseFactoryInterface $responseFactory,
    \Yiisoft\User\Login\Cookie\CookieLogin $cookieLogin,
    \Yiisoft\User\CurrentUser $currentUser
): \Psr\Http\Message\ResponseInterface {
    $response = $responseFactory->createResponse();
    $body = $request->getParsedBody();
    

>     // Get user identity in based on body contents.

Reviewer: ```suggestion
    // Get user identity based on body content.
```

    
    /** @var \Yiisoft\User\Login\Cookie\CookieLoginIdentityInterface $identity */
    
    if ($currentUser->login($identity) && ($body['rememberMe'] ?? false)) {
        $response = $cookieLogin->addCookie($identity, $response);
    }
    
    return $response;
}
```

In the above `rememberMe` in the request body may come from a "remember me" checkbox in the form. End user decides
if he wants to be logged in automatically. If you do not need the user to be able to choose and want to always use
"remember me", you can enable it via the `forceAddCookie` in `params.php`:

```php
return [
    'yiisoft/user' => [
        'authUrl' => '/login',
        'cookieLogin' => [
            'forceAddCookie' => true,
            'duration' => 'P5D', // 5 days
        ],
    ],
];
```

#### Preventing the substitution of cookies

The login cookie value is stored raw. To prevent the substitution of the cookie value,
you can use a `Yiisoft\Cookies\CookieMiddleware`. For more information, see

> [Yii guide to cookies ](https://github.com/yiisoft/docs/blob/master/guide/en/runtime/cookies.md).

Reviewer: ```suggestion
[Yii guide to cookies](https://github.com/yiisoft/docs/blob/master/guide/en/runtime/cookies.md).
```

> Please note that is mandatory before `Yiisoft\User\Login\Cookie\CookieLoginMiddleware`,

> > in the middleware stack should be located `Yiisoft\Cookies\CookieMiddleware`.

Reviewer: Wording should be adjusted.

> You can see an example of all the above described use of the login functionality using cookies
> in the [yiisoft/demo](https://github.com/yiisoft/demo).

Reviewer: ```suggestion
You can find examples of the above features in the [yiisoft/demo](https://github.com/yiisoft/demo).
```

 */
 
$currentUser = $currentUser->withAccessChecker($accessChecker);
```

To perform the check, use `can()` method:

```php
// The name of the permission (e.g. "edit post") that needs access check.
$permissionName = 'edit-post'; // Required.

// Name-value pairs that would be passed to the rules associated with the roles and permissions assigned to the user.
$params = ['postId' => 42]; // Optional. Default is empty array.

if ($currentUser->can($permissionName, $params)) {
    // Some actions
}
```

Note that in case access checker is not provided via `withAccessChecker()` method, `can()` will always return `false`.

#### Session usage

The current user can store user ID and authentication timeouts for auto-login in the session.
To do this, you need to provide a session (see [yiisoft/session](https://github.com/yiisoft/session)) instance:

```php
/** 
 * @var \Yiisoft\Session\SessionInterface $session
 */
 
$currentUser = $currentUser->withSession($session);
```

You can set timeout (number of seconds), during which the user will be logged
out automatically in case of remaining inactive:

```php
$currentUser = $currentUser->withAuthTimeout(3600);
```

> As well as an absolute timeout (number of seconds), during which the user will be logged
> out automatically regardless of activity:

Reviewer: ```suggestion
Also, an absolute timeout (number of seconds) could be used. The user will be logged
out automatically regardless of activity:
```

```php
$currentUser = $currentUser->withAbsoluteAuthTimeout(3600);
```

> By default, timeouts are not used, the user will be logged out after the current session expires.

Reviewer: ```suggestion
By default, timeouts are not used, so the user will be logged out after the current session expires.
```

#### Using with event loop

The `Yiisoft\User\CurrentUser` instance is stateful, so when you build long-running applications
with tools like [Swoole](https://www.swoole.co.uk/) or [RoadRunner](https://roadrunner.dev/) you should reset
the state at every request. For this purpose, you can use the `clear()` method.

### Auto login through identity from request attribute

> For auto login, you can use the `Yiisoft\User\Login\LoginMiddleware`. This middleware  automatically logs user

Reviewer: ```suggestion
For auto login, you can use the `Yiisoft\User\Login\LoginMiddleware`. This middleware automatically logs user
```
